File: labs/module8/Fortran/routine/distance_map_acc.f90
module distance_map_acc_m
   use  dist_cuda_m
   use cudafor

contains
   subroutine distance_map_acc(A, B, C, m, n)
      implicit none
      real, intent(in) :: A(:)
      real, intent(in) :: B(:)
      real, intent(inout) :: C(:,:)
      integer, intent(in) :: m, n
         
      integer :: i, j

!$acc parallel loop copyin(A, B) copyout(C)
      do j = 1,m
!$acc loop
         do i = 1,n
            C(i,j) = dist_cuda(A(j), B(i) )
         end do
      end do
   end subroutine distance_map_acc
end module distance_map_acc_m

File: labs/module8/C/routine/distance_map_acc.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

extern float dist_cuda(float, float);

#pragma acc routine seq
float dist_acc(float a, float b) {
	return sqrt(a*a + b*b);
}

void distance_map_acc(float *A, float *B, float *C, int m, int n)
{
#pragma acc parallel loop copyin(A[0:m],B[0:n]) \
copyout(C[0:m*n])
	for( int i = 0; i < m; i++ ) {
#pragma acc loop
		for( int j = 0; j < n; j++ ) {
			C[i*n+j] = dist_cuda(A[i], B[j]);
		}
	}
	
}
